x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityStatusChangeListener.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.support;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.license.LicenseStateListener;
+import org.elasticsearch.license.XPackLicenseState;
+
+import java.util.Objects;
+
+/**
+ * A listener for license state changes that provides log messages when a license change
+ * causes security to switch between enable and disabled (or vice versa).
+ */
+public class SecurityStatusChangeListener implements LicenseStateListener {
+
+    private final Logger logger;
+    private final XPackLicenseState licenseState;
+    private Boolean securityEnabled;
+
+    public SecurityStatusChangeListener(XPackLicenseState licenseState) {
+        this.logger = LogManager.getLogger(getClass());
+        this.licenseState = licenseState;
+        this.securityEnabled = null;
+    }
+
+    /**
+     * This listener will not be registered if security has been explicitly disabled, so we only need to account for dynamic changes due
+     * to changes in the applied license.
+     */
+    @Override
+    public synchronized void licenseStateChanged() {
+        final boolean newState = licenseState.isSecurityAvailable() && licenseState.isSecurityDisabledByLicenseDefaults() == false;
+        // old state might be null (undefined) so do Object comparison
+        if (Objects.equals(newState, securityEnabled) == false) {
```

Comment:
If `securityEnabled` is null  and `newState` is false, won't this produce a misleading log entry saying
```
disabling security for licence [X]
```

?

Reply:
Why do you say that is misleading?

Reply:
Because we transition from a "Security not enabled" state to a "Security not enabled" state and we log a message saying "disabling security for license [x]" that sounds like it indicates that it was the change in license that disabled security - when it might well be the existing configuration that disables this.

Reply:
I reject all of your assertions :)

In my mind we transition from an uninitialised state (for which security is partially enabled - this is how uninitialised nodes can join a cluster with security enabled) to a "security not enabled" state.

Secondly, I intentionally wrote the log message to strike a balance between vague and informative. It simply states whether security is enabled or not, and what the license is, without explicitly stating anything more than that.

Thirdly, per the method doc, if existing configuration explicitly disables security than this listener is never registered, and we don't log anything, so the message is always indicating that a dynamic decision was made based on the license.

The case where it could be argued to be misleading is if security is explicitly enabled, in which case _every_ license type will trigger a "enabling security for license [x]" message, when clearly the license mode was irrelevant (but the message is still technically true, we are enabling security for your license whether it be basic through to platinum).

Reply:
In seriousness, I do think we want to print a message on startup because it's nice to have a clear confirmation as to whether security is enabled.
I'm happy to change the wording if there is a version that you feel would be less misleading.

Reply:
I see your point(s). I don't have much to offer in terms of less ambiguity. Maybe passive voice instead of gerund, in order to focus more on the license state for the cluster instead of the act of changing it ?

```
logger.info("{} security for license [{}]", newState ? "is enabled" : "is disabled", licenseState.getOperationMode());
```

but I'm not strongly positive this is so much better.

Reply:
How about this?

> Active license is now [{}] ; Security is {enabled|disabled}

Reply:
WFM !